python/cuml/preprocessing/encoders.py

```diff
@@ -0,0 +1,291 @@
+        self._fitted = False
+        self.drop_idx_ = None
+        self._encoders = None
+        if sparse and np.dtype(dtype) not in ['f', 'd', 'F', 'D']:
+            raise ValueError('Only float32, float64, complex64 and complex128 '
+                             'are supported when using sparse')
+
+    def _validate_keywords(self):
+        if self.handle_unknown not in ('error', 'ignore'):
+            msg = ("handle_unknown should be either 'error' or 'ignore', "
+                   "got {0}.".format(self.handle_unknown))
+            raise ValueError(msg)
+        # If we have both dropped columns and ignored unknown
+        # values, there will be ambiguous cells. This creates difficulties
+        # in interpreting the model.
+        if self.drop is not None and self.handle_unknown != 'error':
+            raise ValueError(
+                "`handle_unknown` must be 'error' when the drop parameter is "
+                "specified, as both would create categories that are all "
+                "zero.")
+
+    def _check_is_fitted(self):
+        if not self._fitted:
+            raise RuntimeError("Model must first be .fit()")
+
+    def _compute_drop_idx(self):
+        if self.drop is None:
+            return None
+        elif isinstance(self.drop, str) and self.drop == 'first':
+            return {feature: cp.array([0])
+                    for feature in self._encoders.keys()}
+        elif isinstance(self.drop, dict):
+            if len(self.drop.keys()) != len(self._encoders):
+                msg = ("`drop` should have as many columns as the number "
+                       "of features ({}), got {}")
+                raise ValueError(msg.format(len(self._encoders),
+                                            len(self.drop.keys())))
+            drop_idx = dict()
+            for feature in self.drop.keys():
+                cats = self._encoders[feature].classes_
+                self.drop[feature] = Series(self.drop[feature])
+                if len(self.drop[feature]) != 1:
+                    msg = ("Trying to drop multiple values for feature {}, "
+                           "this is not supported.").format(feature)
+                    # Dropping multiple values actually works except in inverse
+                    # transform where there is no way to know which categories
+                    # where present before one hot encoding if multiples
+                    # categories where dropped.
+                    raise ValueError(msg)
+                if not self.drop[feature].isin(cats).all():
+                    msg = ("Some categories for feature {} were supposed "
+                           "to be dropped, but were not found in the encoder "
+                           "categories.".format(feature))
+                    raise ValueError(msg)
+                cats = Series(cats)
+                idx = cats.isin(self.drop[feature])
+                drop_idx[feature] = cp.asarray(cats[idx].index)
```

Comment:
Same comment about index on GPU

Reply:
In this case, `cats[idx].index` is a `cudf.GenericIndex` which is already on the GPU. There is no copy with `cp.asarray` in that case. I don't think there is benefit in putting it explicitly on the CPU. Reference:
```python
import cudf
import cupy as cp

def get_ptr(ary):
    try:
        pointer, _ = ary.__cuda_array_interface__['data']
    except AttributeError:
        pointer, _ = ary.__array_interface__['data']
    return pointer

gen = cudf.core.GenericIndex([1])
get_ptr(gen)  # 140446801952768
get_ptr(gen.to_array())  # 94097525112496
get_ptr(cp.asarray(gen))  # 140446801952768
```

Reply:
I see your point above with the if-else. Is there a real need to convert to cp array since you are already working with cudf objects so the index itself should work, right?

Reply:
It doesn't work because I use it to index a cupy array later and cupy doesn't accept GenericIndex as input:
https://github.com/rapidsai/cuml/blob/8ea322253ca40cf5637a7c128e1513af54ea5d1f/python/cuml/preprocessing/encoders.py#L210-L212